forge-game/src/main/java/forge/game/staticability/StaticAbilityDisableTriggers.java

```diff
@@ -0,0 +1,129 @@
+        }
+
+        for (final Card ca : cardList) {
+            for (final StaticAbility stAb : ca.getStaticAbilities()) {
+                if (!stAb.getParam("Mode").equals(MODE) || stAb.isSuppressed() || !stAb.checkConditions()) {
+                    continue;
+                }
+
+                if (isDisabled(stAb, triggerType, regtrig, runParams)) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public static boolean isDisabled(final StaticAbility stAb, final TriggerType triggerType, final Trigger regtrig, final Map<AbilityKey, Object> runParams) {
+        // CR 603.2e
+        if (stAb.hasParam("ValidCard") && regtrig.getSpawningAbility() != null) {
+            return false;
+        }
+
+        if (!stAb.matchesValidParam("ValidCard", regtrig.getHostCard())) {
+            return false;
+        }
+
+        if (stAb.hasParam("ValidMode")) {
+            if (!ArrayUtils.contains(stAb.getParam("ValidMode").split(","), regtrig.getMode().toString())) {
+                return false;
+            }
+        }
+
+        if (triggerType.equals(TriggerType.ChangesZone)) {
+            // Cause of the trigger – the card changing zones
+            if (!stAb.matchesValidParam("ValidCause", runParams.get(AbilityKey.Card))) {
+                return false;
+            }
+            if (!stAb.matchesValidParam("Destination", runParams.get(AbilityKey.Destination))) {
+                return false;
+            }
+            if (!stAb.matchesValidParam("Origin", runParams.get(AbilityKey.Origin))) {
+                return false;
+            }
+            if ("Graveyard".equals(runParams.get(AbilityKey.Destination))
+                    && "Battlefield".equals(runParams.get(AbilityKey.Origin))) {
+                // Allow triggered ability of a dying creature that triggers
+                // only when that creature is put into a graveyard from anywhere
+                if ("Card.Self".equals(regtrig.getParam("ValidCard"))
+                        && (!regtrig.hasParam("Origin") || "Any".equals(regtrig.getParam("Origin")))) {
+                    return false;
+                }
+            }
+        } else if (triggerType.equals(TriggerType.ChangesZoneAll)) {
+            final String origin = stAb.getParam("Origin");
+            final String destination = stAb.getParam("Destination");
+            final CardZoneTable table = (CardZoneTable) runParams.get(AbilityKey.Cards);
+
+            // purge all forbidden causes from table
+            CardZoneTable filtered = new CardZoneTable();
+            boolean possiblyDisabled = false;
+            for (Cell<ZoneType, ZoneType, CardCollection> cell : table.cellSet()) {
+                // this currently assumes the table will not contain multiple destinations
+                // however with some effects (e.g. Goblin Welder) that should indeed be the case
```

Comment:
can we try to fix Goblin Welder with that MR too or would that be too complex?

if not we should try to make a follow up MR for this

Reply:
whats the current stand with Goblin Welder? does it work or do we need extra MR?

Reply:
Turned into #2209